orttraining/orttraining/core/optimizer/memory_optimizer/common.cc

```diff
@@ -0,0 +1,312 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include <charconv>
+
+#include "orttraining/core/optimizer/memory_optimizer/common.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph_viewer.h"
+#include "core/framework/tensorprotoutils.h"
+
+#include "core/common/string_utils.h"
+
+namespace onnxruntime::optimizer::memory_optimizer {
+
+namespace {
+
+/**
+ * @brief Prepare info including activation usage, node usage in fw and bw.
+ *
+ * @param graph Graph to iterate.
+ * @param boundary_op_order_in_topological_sort index of the boundary op between fw and bw.
+ * @param node_index_to_its_order_in_topological_sort_map The mapping of node index to its order in topological sort.
+ * @param fw_op_output_arg_used_map Collected activation usage mapping.
+ *   - key: node arg name
+ *   - value: a pair of bool, representing whether the activation is used by forward nodes or by backward nodes.
+ * @param is_forward_nodes Collected node is forward pass op mapping.
+ */
+void GetForwardOutputUsageMap(const Graph& graph,
+                              const ptrdiff_t boundary_op_order_in_topological_sort,
+                              const InlinedHashMap<NodeIndex, size_t>&
+                                  node_index_to_its_order_in_topological_sort_map,
+                              InlinedHashMap<std::string, std::pair<bool, bool>>& fw_op_output_arg_used_map,
+                              InlinedHashMap<const Node*, bool>& is_forward_nodes) {
+  fw_op_output_arg_used_map.clear();
+  ORT_ENFORCE(boundary_op_order_in_topological_sort >= 0);
+  GraphViewer graph_viewer(graph);
+  const auto& node_ids = graph_viewer.GetNodesInTopologicalOrder();
+  is_forward_nodes.reserve(node_ids.size());
+
+  auto is_forward_pass_operator = [](ptrdiff_t op_order_in_topological_sort,
+                                     ptrdiff_t boundary_op_order_in_topological_sort) -> bool {
+    return op_order_in_topological_sort <= boundary_op_order_in_topological_sort;
+  };
+
+  for (size_t i = 0; i < node_ids.size(); ++i) {
+    const Node* p_node = graph.GetNode(node_ids[i]);
+    if (p_node == nullptr /* skip removed nodes*/) {
+      continue;
+    }
+
+    const Node& node = *p_node;
+
+    bool is_forward_op = is_forward_pass_operator(static_cast<ptrdiff_t>(i), boundary_op_order_in_topological_sort);
+    if (!is_forward_op) {
+      is_forward_nodes[p_node] = false;
+      continue;
+    }
+
+    is_forward_nodes[p_node] = true;
+
+    for (auto& output_arg : node.OutputDefs()) {
+      bool used_in_fw = false;
+      bool used_in_bw = false;
+      for (auto& consumer_node : graph.GetConsumerNodes(output_arg->Name())) {
+        size_t consumer_node_index_in_topological_order =
+            node_index_to_its_order_in_topological_sort_map.at(consumer_node->Index());
```

Comment:
>node_index_to_its_order_in_topological_sort_map.at(consumer_node->Index()); [](http://example.com/codeflow?start=12&length=75)

This will throw std::out_of_range exception with not much information to debug this if a customer files an issue. Why not check for existence yourself and throw a meaningful message with ORT_ENFORCE?